### PR TITLE
feat(object): add object lock configuration resource

### LINKED ISSUE #1467: Resource Object Storage: Setting up object lock
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The object lock API feature allows users to lock objects and prevent them from being deleted or overwritten. Objects can be put on lock for a specific amount of time or indefinitely. The lock period is defined by the user.

### New or Affected Resource(s)

* scaleway_object_bucke_lock_configuration

### Potential Terraform Configuration

```hcl
resource "scaleway_object_bucket" "example" {
  name = "mybucket"

  object_lock_enabled = true
}

resource "scaleway_object_bucke_lock_configuration" "example" {
  bucket = scaleway_object_bucket.example.name

  rule {
    default_retention {
      mode = "COMPLIANCE"
      days = 5
    }
  }
}
```

### References
https://www.scaleway.com/en/docs/storage/object/api-cli/object-lock/

* #0000

### PR DESCRIPTION
**Pull request by quantumsheep**
Resolves #1467.

### CODE REVIEW

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+										Required:     true,
+										ValidateFunc: validation.StringInSlice([]string{"GOVERNANCE", "COMPLIANCE"}, false),
+										Description:  "The default Object Lock retention mode you want to apply to new objects placed in the specified bucket.",
+									},
+									"days": {
+										Type:        schema.TypeInt,
+										Optional:    true,
+										Description: "The number of days that you want to specify for the default retention period.",
+									},
+									"years": {
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:65]: There is a conflict with days and years in the schema

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+	}
+
+	bucket := expandID(d.Get("bucket").(string))
+
+	lockConfig := &s3.ObjectLockConfiguration{
+		ObjectLockEnabled: aws.String("Enabled"),
+		Rule:              expandBucketLockConfigurationRule(d.Get("rule").([]interface{})),
+	}
+
+	_, err = conn.ListObjectsWithContext(ctx, &s3.ListObjectsInput{
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:94]: Not necessary

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+		return diag.FromErr(err)
+	}
+
+	input := &s3.GetObjectLockConfigurationInput{
+		Bucket: aws.String(bucket),
+	}
+
+	// expectedBucketOwner and routing not supported
+
+	_, err = conn.ListObjectsWithContext(ctx, &s3.ListObjectsInput{
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:136]: Not necessary

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+			tflog.Error(ctx, fmt.Sprintf("Bucket %q was not found - removing from state!", bucket))
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(fmt.Errorf("couldn't read bucket: %s", err))
+	}
+
+	output, err := conn.GetObjectLockConfigurationWithContext(ctx, input)
+	if !d.IsNewResource() && ErrCodeEquals(err, s3.ErrCodeNoSuchBucket) {
+		tflog.Debug(ctx, fmt.Sprintf("[WARN] Object Bucket Lock Configuration (%s) not found, removing from state", d.Id()))
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:150]: Not a debug but a warning

File: docs/resources/object_bucket_object_lock_configuration.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+The following arguments are supported:
+
+- `bucket` - (Required, Forces new resource) The name of the bucket.
+
+- `rule` - (Optional) Specifies the Object Lock rule for the specified object.
+
+    - `default_retention` - (Required) The default retention for the lock.
+
+        - `mode` - (Required) The default Object Lock retention mode you want to apply to new objects placed in the specified bucket.
```
Comment (Monitob) [inline docs/resources/object_bucket_object_lock_configuration.md:52]: Add the supported modes

**Review by Monitob — CHANGES_REQUESTED** [review]

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 6]
```diff
@@ -59,13 +59,13 @@ func resourceObjectLockConfiguration() *schema.Resource {
 										Type:          schema.TypeInt,
 										Optional:      true,
 										Description:   "The number of days that you want to specify for the default retention period.",
-										ConflictsWith: []string{"years"},
+										ConflictsWith: []string{"rule.default_retention.years"},
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration.go:62]:
```suggestion
										ConflictsWith: []string{"rule.0.default_retention.0.years"},
```

File: docs/resources/object_bucket_object_lock_configuration.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+The following arguments are supported:
+
+- `bucket` - (Required, Forces new resource) The name of the bucket.
+
+- `rule` - (Optional) Specifies the Object Lock rule for the specified object.
+
+    - `default_retention` - (Required) The default retention for the lock.
+
+        - `mode` - (Required) The default Object Lock retention mode you want to apply to new objects placed in the specified bucket. Valid values are `GOVERNANCE` or `COMPLIANCE`.
```
Comment (remyleone) [inline docs/resources/object_bucket_object_lock_configuration.md:52]: What does GOVERNANCE and CONPLIANCE mean in this context?

Comment (remyleone) [reply docs/resources/object_bucket_object_lock_configuration.md:52]: Link to https://www.scaleway.com/en/docs/storage/object/api-cli/object-lock/#retention-modes

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+const (
+	objectBucketLockConfigurationRetry = 2 * time.Minute
```
Comment (remyleone) [inline scaleway/resource_object_bucket_lock_configuration.go:17]: it seems like a lot

File: scaleway/resource_object_bucket.go [hunk 9]
```diff
@@ -431,6 +440,31 @@ func resourceScalewayObjectBucketRead(ctx context.Context, d *schema.ResourceDat
...
+				tflog.Error(ctx, fmt.Sprintf("Bucket %q was not found - removing from state!", bucketName))
+				d.SetId("")
+				return nil
+			case "ObjectLockConfigurationNotFoundError":
+				_ = d.Set("object_lock_enabled", false)
+			default:
+				return diag.FromErr(fmt.Errorf("couldn't read bucket: %s", err))
+			}
+		} else {
+			return diag.FromErr(fmt.Errorf("couldn't read bucket: %s", err))
```
Comment (Codelax) [inline scaleway/resource_object_bucket.go:462]: use `isS3Error` and define `"ObjectLockConfigurationNotFoundError"` in a constant in `errors_object.go`

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	input := &s3.GetObjectLockConfigurationInput{
+		Bucket: aws.String(bucket),
+	}
+
+	output, err := conn.GetObjectLockConfigurationWithContext(ctx, input)
+	if !d.IsNewResource() && ErrCodeEquals(err, s3.ErrCodeNoSuchBucket) {
```
Comment (Codelax) [inline scaleway/resource_object_bucket_lock_configuration.go:116]: ErrCodeEquals works with scaleway sdk errors, use `isS3Error`

File: scaleway/resource_object_bucket_lock_configuration.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+	input := &s3.PutObjectLockConfigurationInput{
+		Bucket: aws.String(bucket),
+		ObjectLockConfiguration: &s3.ObjectLockConfiguration{
+			ObjectLockEnabled: aws.String(s3.ObjectLockEnabledEnabled),
+		},
+	}
+
+	_, err = conn.PutObjectLockConfigurationWithContext(ctx, input)
+
+	if ErrCodeEquals(err, s3.ErrCodeNoSuchBucket) {
```
Comment (Codelax) [inline scaleway/resource_object_bucket_lock_configuration.go:178]: use object helpers for s3 errors

File: scaleway/resource_object_bucket_lock_configuration_test.go [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+func TestAccObjectBucketLockConfiguration_update(t *testing.T) {
+	if !*UpdateCassettes {
+		t.Skip("Skipping ObjectStorage test as this kind of resource can't be deleted before 24h")
+	}
+	rName := sdkacctest.RandomWithPrefix(LockResourcePrefix)
+	resourceName := lockResourceTestName
+
+	tt := NewTestTools(t)
```
Comment (Codelax) [inline scaleway/resource_object_bucket_lock_configuration_test.go:85]: Do not skip tests with cassettes and init correctly TestTools, defer Cleanup is missing

File: scaleway/resource_object_bucket_lock_configuration_test.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+const (
+	LockResourcePrefix   = "tf-acc-test"
+	lockResourceTestName = "scaleway_object_bucket_lock_configuration.test"
+)
+
+func TestAccObjectBucketLockConfiguration_basic(t *testing.T) {
```
Comment (Codelax) [inline scaleway/resource_object_bucket_lock_configuration_test.go:19]: Tests naming does not follow naming of others. Cassettes don't have resource name.

**Review by Codelax — DISMISSED** [review]

File: scaleway/resource_object_bucket_lock_configuration_test.go [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+						}
+					}
+				`, rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckBucketLockConfigurationExists(tt, resourceName),
+					resource.TestCheckResourceAttrPair(resourceName, "bucket", "scaleway_object_bucket.test", "name"),
+					resource.TestCheckResourceAttr(resourceName, "rule.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.default_retention.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.default_retention.0.mode", "GOVERNANCE"),
+					resource.TestCheckResourceAttr(resourceName, "rule.0.default_retention.0.days", "1"),
```
Comment (Monitob) [inline scaleway/resource_object_bucket_lock_configuration_test.go:64]: Please add a test with years

File: scaleway/resource_object_bucket.go [hunk 15]
```diff
@@ -36,6 +36,13 @@ func resourceScalewayObjectBucket() *schema.Resource {
 				ForceNew:    true,
 				Description: "The name of the bucket",
 			},
+			"object_lock_enabled": {
```
Comment (Monitob) [inline scaleway/resource_object_bucket.go:39]: Please add this on the bucket documentation

**Review by Monitob — CHANGES_REQUESTED** [review]

**Pull request merged**